ccl/sqlproxyccl: improve connection logging behavior

jaylim-crl opened the pull request:
### ccl/sqlproxyccl: improve connection logging behavior

Previously, the connection logging behavior in the proxy had several issues that led to unnecessary log spam:

1. **Excessive "registerAssignment" and "unregisterAssignment" logs**: These logs were emitted for every connection attempt (including migration) and were not useful under normal operation.
2. **Redundant error logging**: Most errors were logged twice -- once in the `handle` method, and again when it returns.
3. **Unfiltered error hints**: User-facing errors containing hints were logged line by line, cluttering the logs.
4. **Lack of context in error logs**: Errors logged from the proxy lacked tenant and cluster context, which made troubleshooting more difficult.

This commit addresses these issues as follows:

1. **Reduced logging**: "registerAssignment" and "unregisterAssignment" logs are now only shown with vmodule logging enabled.
2. **Error logging improvements**: `handle` no longer logs errors (with the exception of some cases); the caller is now responsible for logging. Additionally, errors with hints are no longer logged, only the main error is recorded. When those errors are logged, they will now include the tenant and cluster information where possible.

No release note as this is an internal change.

Epic: none

Release note: None

### ccl/sqlproxyccl: rate limit throttler errors from being logged

Previously, each request refused by the throttler would result in a "throttler refused connection" message being logged, generating a log entry for every rejected request.

The throttler service is responsible for rate limiting invalid login attempts from IP addresses, and in practice, it can generate a high volume of such traffic. To address this, errors are now rate limited in the logs to occur once every 5 minutes per (IP, tenant) pair, ensuring that only one log entry is generated within that time frame.

This change is internal, so no release note is required.

Epic: none

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: improve connection logging behavior

Previously, the connection logging behavior in the proxy had several issues
that led to unnecessary log spam:

1. **Excessive "registerAssignment" and "unregisterAssignment" logs**: These
   logs were emitted for every connection attempt (including migration) and
   were not useful under normal operation.
2. **Redundant error logging**: Most errors were logged twice -- once in the
   `handle` method, and again when it returns.
3. **Unfiltered error hints**: User-facing errors containing hints were logged
   line by line, cluttering the logs.
4. **Lack of context in error logs**: Errors logged from the proxy lacked
   tenant and cluster context, which made troubleshooting more difficult.

This commit addresses these issues as follows:

1. Reduced logging: "registerAssignment" and "unregisterAssignment" logs are
   now only shown with vmodule logging enabled.
2. Error logging improvements: `handle` no longer logs errors (with the
   exception of some cases); the caller is now responsible for logging.
   Additionally, errors with hints are no longer logged, only the main error is
   recorded. When those errors are logged, they will now include the tenant
   and cluster information where possible.

No release note as this is an internal change.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134613)
<!-- Reviewable:end -->

jaylim-crl committed:
ccl/sqlproxyccl: rate limit throttler errors

Previously, each request refused by the throttler would result in a
"throttler refused connection" message being logged, generating a log entry
for every rejected request.

The throttler service is responsible for rate limiting invalid login attempts
from IP addresses, and in practice, it can generate a high volume of such
traffic. To address this, errors are now rate limited in the logs to occur
once every 5 minutes per (IP, tenant) pair, ensuring that only one log entry
is generated within that time frame.

This change is internal, so no release note is required.

Epic: none

Release note: None

jaylim-crl commented on pkg/ccl/sqlproxyccl/throttler/local.go:
```diff
@@ -86,18 +88,26 @@ func (s *localService) lockedInsertThrottle(connection ConnectionTags) *throttle
 	now := s.clock()
 	throttle := s.lockedGetThrottle(connection)
 	if throttle != nil && throttle.isThrottled(now) {
+		if throttle.everyLog.ShouldLog() {
+			// ctx should include logtags about the connection.
+			log.Error(ctx, "throttler refused connection due to too many failed authentication attempts")
+		}
```
If we didn't want the service to probe into the internals of the throttle struct (e.g. everyLog), this could be its own method on throttle (e.g. reportThrottled), but I don't feel strongly.

davidwding commented on pkg/ccl/sqlproxyccl/balancer/conn_tracker.go:
```diff
@@ -49,7 +54,7 @@ func NewConnTracker(
 	}
 
-	t := &ConnTracker{timeSource: timeSource}
+	t := &ConnTracker{timeSource: timeSource, verboseLogging: log.V(2)}
```
What is log level 2 and how did you pick it? Are there corresponding constants like error/warn/info/etc.?

jaylim-crl replied:
Unfortunately, I don't think we have any rule of thumb around vmodule logging levels. See this internal Slack thread: https://cockroachlabs.slack.com/archives/C9TGBJB44/p1614100926026700. Each vmodule logging level seems to be specific to the file itself. I picked 2 somewhat arbitrarily; it's a middle value, and opens up opportunities for something less/more verbose as well.

For this case, if we wanted to display logs from the connection tracker, we would start the proxy with the following flag: `--vmodule=conn_tracker=2` (i.e. `--vmodule=FILE=LEVEL,FILE2=LEVEL2,...`).

davidwding commented on pkg/ccl/sqlproxyccl/proxy_handler.go:
```diff
@@ -372,34 +369,44 @@ func (handler *proxyHandler) handle(
+		return errors.Wrap(err, "extracting cluster identifier")
+	}
+
+	// Validate the incoming connection and ensure that the cluster name
```
Why did this block of code move?

jaylim-crl replied:
I'd like to ensure that we only add logtags after validating the connection. If we left the validation in its original location, we may end up having a log line with a mismatched cluster name and tenant ID, leading to confusions during debugging. See the block that comes after the validation:

```go
// Only add logtags after validating the connection. If the connection isn't
// validated, clusterName may not match the tenant ID, and this could cause
// confusion when analyzing logs.
ctx = logtags.AddTag(ctx, "cluster", clusterName)
ctx = logtags.AddTag(ctx, "tenant", tenID)
```

The proxy uses the tenant ID to spin up a new SQL server.  Back then, there were concerns that users could iterate through `--cluster=random-name-1` to `--cluster=random-name-N`, causing the operator to inadvertently provision resources for those tenants, and leading to potential resource exhaustion on the host. To address that, we implemented a validation check to ensure that "random-name" matches the actual cluster name associated with that tenant ID (as seen in the DnsPrefix field in the CrdbTenant spec, or the tenants table in CC). This validation guarantees that the user connecting to tenant N has _some_ knowledge of the cluster, beyond just the tenant ID, and prevents random or malicious requests from triggering resource allocation.

davidwding commented on pkg/ccl/sqlproxyccl/throttler/local.go:
```diff
@@ -72,6 +72,8 @@ func NewLocalService(opts ...LocalOption) Service {
 	return s
 }
 
+var _ Service = (*localService)(nil)
```
What does this line do?

jaylim-crl replied:
This ensures that `localService` implements the `Service` interface at compile time. This is the same as doing:

```go
var _ Service = &localService{}
```

We use this pattern in CC as well.

davidwding commented on pkg/ccl/sqlproxyccl/server.go:
```diff
@@ -268,12 +268,38 @@ func (s *Server) serve(ctx context.Context, ln net.Listener, requireProxyProtoco
+				for key, value := range reqTags {
+					ctx = logtags.AddTag(ctx, key, value)
+				}
+				// log.Infof automatically prints hints (one per line) that are
```
Can we also change the behavior of `log.Infof` or is that a much larger task?

jaylim-crl replied:
The `log` package (https://github.com/cockroachdb/cockroach/tree/master/pkg/util/log) is utilized extensively throughout the CRDB codebase, and I'd imagine there could be a scenario where we actually want hints. For the proxy's use case, I think the right thing to do here is to avoid returning user-facing error messages from `handle`. This would require a deeper audit, and some refactoring of the logic and tests. For now, this approach works to address the unnecessary log spam issue.

davidwding reviewed (commented):
Some basic questions, this is all new code to me

jaylim-crl commented:
Thanks for the reviews!

davidwding reviewed (approved):
Thanks for all the explanations!
